Store cart ID in cookies

scottdixon opened the pull request:
Hey @lordofthecactus! Let me know what you think of this approach.

### WHY are these changes introduced?

When switching from Shopify's Online Store to Hydrogen, existing customer carts are lost unless the merchant updates their liquid theme to use the Storefront API.

Soon, liquid/Storefront API carts will be interoperable - carts created on a traditional liquid storefront can be managed via the Storefront API and vice versa.

This change leverages the new functionality and shows an example of how to switch between Online Store/Hydrogen without losing carts.

### WHAT is this pull request doing?

Updates the Demo Store template to use the same `cart` cookie [set by the traditional Online Store channel](https://screenshot.click/27-52-gldtj-mzwjg.png).

### HOW to test your changes?

1. Enable `cart_repository_online_store_path` on your test store.
2. Create a cart using the traditional Online Store and copy the cart ID from the `cart` cookie.
3. Running Demo Store locally, create the same `cart` cookie.
4. Switch between Hydrogen/Online Store and the cart should remain in sync.

Feel free to use my test store:

```
PUBLIC_STORE_DOMAIN="test-shop-987.myshopify.com"
PUBLIC_STOREFRONT_API_TOKEN="b5f455ac1cea1522791da1ad0e053559"
```

#### Post-merge steps

Will remove most of these docs: https://shopify.dev/docs/custom-storefronts/oxygen/migrate/share-carts

#### Checklist

- [x] I've read the [Contributing Guidelines](CONTRIBUTING.md)
- [x] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [x] I've added a [changeset](CONTRIBUTING.md#changesets) if this PR contains user-facing or noteworthy changes
- [ ] I've added [tests](CONTRIBUTING.md#testing) to cover my changes
- [ ] I've added or updated the documentation

lordofthecactus reviewed (changes requested):
Hi @scottdixon 👋 PR works well! I have been 🎩 (tophatting) with your demo store and I could go back and forth the online storefront and hydrogen 🔥 

One request before merging!

I also did a deep dive on sessions in remix and made sure:
* Using `request.headers` is alright
* The `Set-Cookie` works as intended and does not affect other parts of the app. Only if we have another `Set-Cookie` with the same cookie name would this get overwritten.

Also:
* fixed the types
* rebased from latests

lordofthecactus commented on templates/demo-store/app/routes/($lang)/cart.tsx:
```diff
@@ -71,6 +69,7 @@ export async function action({request, context}: ActionArgs) {
 
       break;
     case CartAction.REMOVE_FROM_CART:
+      invariant(cartId, 'Missing cartId');
```
@scottdixon cart was complaining here because when you parsed the cookie with `getCartId` the function would have `string | undefined` union. 

Then some of the actions below required to have a string. So instead I just did a check with `invariant` if the cartId actually exists.

lordofthecactus commented:
@scottdixon I'm also wondering if we should update to add some text here https://shopify.dev/docs/custom-storefronts/hydrogen/cart/add#persist-and-return-the-cart about this. 
Maybe some sort of note, that if cart wants to be shared with online storefront, to setup a plan cookie, not within session

scottdixon commented:
Thanks so much @lordofthecactus! I've updated changeset and added a note to the cart docs [as part of this PR](https://github.com/Shopify/shopify-dev/pull/32466) which will roll out once the feature is live. LMK if I've missed anything 🙏
